Este notebook se ejecuta en Google Colab, pues la intención es simplemente extraer nuestros datos de Kaggle.

Trabajaremos con los datos IEEE-CIS Fraud Detection, sin embargo, la intención no es trabajar con todos, pues es una base de datos grande y la consecuencia de trabajar con todos será obtener modelos tambien muy pesados. Por lo que haremos una muestra de los datos de entrenamiento.

Además aprovecharemos el proceso de la muestra, para incluir los pasos del balance de datos en caso de ser necesario, pues la experiencia nos dice que los datasets de detección de fraude generalmente contienen muchos datos de "No fraude" y muy pocos de "Fraude", esto es una fiel representación de la realidad, pero para entrenar modelos de Machine Learning efectivos, necesitamos darles infromación de amabas clases para que aprendan bien de amabas.   



In [ ]:
# Este paso es necesario para poder extraer el dataset de nuestro inteeres en Kaggle
!pip install kaggle==1.5.12

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 59.0/59.0 kB 4.0 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... done
  Created wheel for kaggle: filename=kaggle-1.5.12-py3-none-any.whl size=73026 sha256=79d459507619f323ffdd6acc20be4404bd2b41253adc1989d7efa3eb29f8f45c
  Stored in directory: /root/.cache/pip/wheels/2e/27/39/f44e52756a6407b444143f233abe9fda0e18a23e8b20e0cd1c
Successfully built kaggle
  Attempting uninstall: kaggle
    Found existing installation: kaggle 1.6.17
    Uninstalling kaggle-1.6.17:
      Successfully uninstalled kaggle-1.6.17


Para poder descargar el archivo nos dirijimos a nuestro perfil de Kaggle y le damos click en "Create New API Token", para que se nos descargue el archivo kaggle.json, que es una clave de API con un token unico que nos identifica. A continuación usamos los siguientes comandos para tener el archivo en el lugar correcto.

In [7]:
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/
!chmod 600 ~/.kaggle/kaggle.json

Una vez hecho eso, proecedemos a descargar nuestros datos.

In [11]:
import pandas as pd
!kaggle competitions download -c ieee-fraud-detection



 98% 116M/118M [00:00<00:00, 189MB/s] 
100% 118M/118M [00:00<00:00, 181MB/s]


Descomprimimos el archivo zip para poder ver los diferentes archivos.

In [13]:
!unzip ieee-fraud-detection.zip

Archive:  ieee-fraud-detection.zip
  inflating: sample_submission.csv   
  inflating: test_identity.csv       
  inflating: test_transaction.csv    
  inflating: train_identity.csv      
  inflating: train_transaction.csv   


A nosostros solo nos interesan los datos de entrenamiento, "train_transaccion.csv" y "train_identity.csv, los datos "test" carecen de la columna "IsFraud", por lo que no los usaremos para nuestra muestra.

In [14]:
train_transaction = pd.read_csv('train_transaction.csv')
train_identity = pd.read_csv('train_identity.csv')
train_df = pd.merge(train_transaction, train_identity, on='TransactionID', how='left')
print(train_df.head()) # Para ver las primeras filas del DataFrame

   TransactionID  isFraud  TransactionDT  TransactionAmt ProductCD  card1  \
0        2987000        0          86400            68.5         W  13926   
1        2987001        0          86401            29.0         W   2755   
2        2987002        0          86469            59.0         W   4663   
3        2987003        0          86499            50.0         W  18132   
4        2987004        0          86506            50.0         H   4497   

   card2  card3       card4  card5  ...                id_31  id_32  \
0    NaN  150.0    discover  142.0  ...                  NaN    NaN   
1  404.0  150.0  mastercard  102.0  ...                  NaN    NaN   
2  490.0  150.0        visa  166.0  ...                  NaN    NaN   
3  567.0  150.0  mastercard  117.0  ...                  NaN    NaN   
4  514.0  150.0  mastercard  102.0  ...  samsung browser 6.2   32.0   

       id_33           id_34  id_35 id_36 id_37  id_38  DeviceType  \
0        NaN             NaN    NaN   Na

Al tener tantas columnas, pandas no nos muestra todas, pero lo podemos solucionar con la siguiente configuración:

In [19]:
pd.set_option('display.max_columns', None)
train_transaction.head()

TransactionID  isFraud  TransactionDT  TransactionAmt ProductCD  card1  \
0        2987000        0          86400            68.5         W  13926   
1        2987001        0          86401            29.0         W   2755   
2        2987002        0          86469            59.0         W   4663   
3        2987003        0          86499            50.0         W  18132   
4        2987004        0          86506            50.0         H   4497   

   card2  card3       card4  card5   card6  addr1  addr2  dist1  dist2  \
0    NaN  150.0    discover  142.0  credit  315.0   87.0   19.0    NaN   
1  404.0  150.0  mastercard  102.0  credit  325.0   87.0    NaN    NaN   
2  490.0  150.0        visa  166.0   debit  330.0   87.0  287.0    NaN   
3  567.0  150.0  mastercard  117.0   debit  476.0   87.0    NaN    NaN   
4  514.0  150.0  mastercard  102.0  credit  420.0   87.0    NaN    NaN   

  P_emaildomain R_emaildomain   C1   C2   C3   C4   C5   C6   C7   C8   C9  \
0           NaN           NaN  1.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
1     gmail.com           NaN  1.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
2   outlook.com           NaN  1.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
3     yahoo.com           NaN  2.0  5.0  0.0  0.0  0.0  4.0  0.0  0.0  1.0   
4     gmail.com           NaN  1.0  1.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0   

   C10  C11  C12   C13  C14     D1     D2    D3    D4   D5  D6  D7  D8  D9  \
0  0.0  2.0  0.0   1.0  1.0   14.0    NaN  13.0   NaN  NaN NaN NaN NaN NaN   
1  0.0  1.0  0.0   1.0  1.0    0.0    NaN   NaN   0.0  NaN NaN NaN NaN NaN   
2  0.0  1.0  0.0   1.0  1.0    0.0    NaN   NaN   0.0  NaN NaN NaN NaN NaN   
3  0.0  1.0  0.0  25.0  1.0  112.0  112.0   0.0  94.0  0.0 NaN NaN NaN NaN   
4  1.0  1.0  0.0   1.0  1.0    0.0    NaN   NaN   NaN  NaN NaN NaN NaN NaN   

    D10    D11  D12  D13  D14    D15   M1   M2   M3   M4   M5   M6   M7   M8  \
0  13.0   13.0  NaN  NaN  NaN    0.0    T    T    T   M2    F    T  NaN  NaN   
1   0.0    NaN  NaN  NaN  NaN    0.0  NaN  NaN  NaN   M0    T    T  NaN  NaN   
2   0.0  315.0  NaN  NaN  NaN  315.0    T    T    T   M0    F    F    F    F   
3  84.0    NaN  NaN  NaN  NaN  111.0  NaN  NaN  NaN   M0    T    F  NaN  NaN   
4   NaN    NaN  NaN  NaN  NaN    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

    M9   V1   V2   V3   V4   V5   V6   V7   V8   V9  V10  V11  V12  V13  V14  \
0  NaN  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0   
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  0.0  0.0  1.0   
2    F  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0   
3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.0  1.0  1.0   
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   V15  V16  V17  V18  V19  V20  V21  V22  V23  V24  V25  V26  V27  V28  V29  \
0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0   
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   V30  V31  V32  V33  V34  V35  V36  V37  V38  V39  V40  V41  V42  V43  V44  \
0  0.0  0.0  0.0  0.0  0.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0   
2  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0   
3  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0   
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   V45  V46  V47  V48  V49  V50  V51  V52  V53  V54  V55  V56  V57  V58  V59  \
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.0  1.0  1.0  1.0  0.0  0.0  0.0   
1  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  

El dataset que nos interesa es el train_df que resulto de unir el dataset de transacciones y el de Identidad, si en nuestra muestra vemos que los datos de identidad no nos aporta mucho podemos eliminar las columnas despues.

In [20]:
train_df.head()

TransactionID  isFraud  TransactionDT  TransactionAmt ProductCD  card1  \
0        2987000        0          86400            68.5         W  13926   
1        2987001        0          86401            29.0         W   2755   
2        2987002        0          86469            59.0         W   4663   
3        2987003        0          86499            50.0         W  18132   
4        2987004        0          86506            50.0         H   4497   

   card2  card3       card4  card5   card6  addr1  addr2  dist1  dist2  \
0    NaN  150.0    discover  142.0  credit  315.0   87.0   19.0    NaN   
1  404.0  150.0  mastercard  102.0  credit  325.0   87.0    NaN    NaN   
2  490.0  150.0        visa  166.0   debit  330.0   87.0  287.0    NaN   
3  567.0  150.0  mastercard  117.0   debit  476.0   87.0    NaN    NaN   
4  514.0  150.0  mastercard  102.0  credit  420.0   87.0    NaN    NaN   

  P_emaildomain R_emaildomain   C1   C2   C3   C4   C5   C6   C7   C8   C9  \
0           NaN           NaN  1.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
1     gmail.com           NaN  1.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  0.0   
2   outlook.com           NaN  1.0  1.0  0.0  0.0  0.0  1.0  0.0  0.0  1.0   
3     yahoo.com           NaN  2.0  5.0  0.0  0.0  0.0  4.0  0.0  0.0  1.0   
4     gmail.com           NaN  1.0  1.0  0.0  0.0  0.0  1.0  0.0  1.0  0.0   

   C10  C11  C12   C13  C14     D1     D2    D3    D4   D5  D6  D7  D8  D9  \
0  0.0  2.0  0.0   1.0  1.0   14.0    NaN  13.0   NaN  NaN NaN NaN NaN NaN   
1  0.0  1.0  0.0   1.0  1.0    0.0    NaN   NaN   0.0  NaN NaN NaN NaN NaN   
2  0.0  1.0  0.0   1.0  1.0    0.0    NaN   NaN   0.0  NaN NaN NaN NaN NaN   
3  0.0  1.0  0.0  25.0  1.0  112.0  112.0   0.0  94.0  0.0 NaN NaN NaN NaN   
4  1.0  1.0  0.0   1.0  1.0    0.0    NaN   NaN   NaN  NaN NaN NaN NaN NaN   

    D10    D11  D12  D13  D14    D15   M1   M2   M3   M4   M5   M6   M7   M8  \
0  13.0   13.0  NaN  NaN  NaN    0.0    T    T    T   M2    F    T  NaN  NaN   
1   0.0    NaN  NaN  NaN  NaN    0.0  NaN  NaN  NaN   M0    T    T  NaN  NaN   
2   0.0  315.0  NaN  NaN  NaN  315.0    T    T    T   M0    F    F    F    F   
3  84.0    NaN  NaN  NaN  NaN  111.0  NaN  NaN  NaN   M0    T    F  NaN  NaN   
4   NaN    NaN  NaN  NaN  NaN    NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

    M9   V1   V2   V3   V4   V5   V6   V7   V8   V9  V10  V11  V12  V13  V14  \
0  NaN  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0   
1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  0.0  0.0  1.0   
2    F  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0   
3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.0  1.0  1.0   
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   V15  V16  V17  V18  V19  V20  V21  V22  V23  V24  V25  V26  V27  V28  V29  \
0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0   
1  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0   
2  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0   
3  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  0.0   
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   V30  V31  V32  V33  V34  V35  V36  V37  V38  V39  V40  V41  V42  V43  V44  \
0  0.0  0.0  0.0  0.0  0.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   
1  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0   
2  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0   
3  0.0  0.0  0.0  0.0  0.0  1.0  1.0  1.0  1.0  0.0  0.0  1.0  0.0  0.0  1.0   
4  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   

   V45  V46  V47  V48  V49  V50  V51  V52  V53  V54  V55  V56  V57  V58  V59  \
0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  1.0  1.0  1.0  1.0  0.0  0.0  0.0   
1  1.0  1.0  1.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  0.0  

In [21]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 590540 entries, 0 to 590539
Columns: 434 entries, TransactionID to DeviceInfo
dtypes: float64(399), int64(4), object(31)
memory usage: 1.9+ GB


Veamos que tan grave es nuestro problema de valores nulos.

In [25]:
# Contamos valores nulos por columna
valores_nulos_por_columna = train_df.isnull().sum()

# Calculamos el porcentaje de valores nulos por columna
porcentaje_nulos_por_columna = (valores_nulos_por_columna / len(train_df)) * 100

# Creamos un DataFrame con los resultados para poder ver todos los datos
resultados_nulos = pd.DataFrame({
    'Columna': valores_nulos_por_columna.index,
    'Valores_Nulos': valores_nulos_por_columna.values,
    'Porcentaje_Nulos': porcentaje_nulos_por_columna.values
})

# Configuramos Pandas para mostrar todas las columnas y filas
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# Imprimimos el DataFrame con los resultados
print("Valores nulos y porcentajes por columna:")
print(resultados_nulos)

# Restablecemos la configuración de Pandas
pd.reset_option('display.max_columns')
pd.reset_option('display.max_rows')

Valores nulos y porcentajes por columna:
            Columna  Valores_Nulos  Porcentaje_Nulos
0     TransactionID              0          0.000000
1           isFraud              0          0.000000
2     TransactionDT              0          0.000000
3    TransactionAmt              0          0.000000
4         ProductCD              0          0.000000
5             card1              0          0.000000
6             card2           8933          1.512683
7             card3           1565          0.265012
8             card4           1577          0.267044
9             card5           4259          0.721204
10            card6           1571          0.266028
11            addr1          65706         11.126427
12            addr2          65706         11.126427
13            dist1         352271         59.652352
14            dist2         552913         93.628374
15    P_emaildomain          94456         15.994852
16    R_emaildomain         453249         76.751617
17   

Veamos ahora que tan desbalanceado esta nuesto DataFrame

In [26]:
# Contamos los valores únicos en la columna 'isFraud'
conteo_isFraud = train_df['isFraud'].value_counts()

print("Conteo de valores en 'isFraud':")
print(conteo_isFraud)

Conteo de valores en 'isFraud':
isFraud
0    569877
1     20663
Name: count, dtype: int64


Ahora podemos ver que contamos con 20663 casos en los que la transaccion fue fraude, en nuestra muestra debemos incluir todos, pues tener el maximo de casos de fraude es nuestro princial objetivo en este momento, la muestra la vamos a complementear con datos aleatiorios dentro de todo el conjunto de "No fraude"; sin embargo, como dijimos al inicio, no nos interesa tener tantos, asi que elegiremos ingresar a nuestra muestra solo el doble de loa datos que si fueron fraude.

Con esto tendremos un DataFrame con 1/3 de datos con fraude y 2/3 con datos de no fraude, aunque seguira habiendo un desbalance, esto no es radical y la mayoría de los modelos podrán trabajar bien con esta proporción. Nos ahorramos tener que recurrir a tecnicas de undersampling o generación de datos sinteticos en los pasos subsecuentes.

In [27]:
# Filtamos los casos de fraude (isFraud = 1)
fraude_df = train_df[train_df['isFraud'] == 1]

# Filtramos los casos de no fraude (isFraud = 0)
no_fraude_df = train_df[train_df['isFraud'] == 0]

# Selecionamos el doble de casos de no fraude respecto a casos de fraude de manera aleatoria
no_fraude_sample = no_fraude_df.sample(n=2 * len(fraude_df), random_state=42)

# Combinamos los casos de fraude y no fraude en un nuevo DataFrame
balanced_df = pd.concat([fraude_df, no_fraude_sample])

# Mezclamos el DataFrame para que los casos estén en orden aleatorio
balanced_df = balanced_df.sample(frac=1, random_state=42).reset_index(drop=True)

# Verificamos el balanceo
print("Conteo de valores en 'isFraud' en el dataset balanceado:")
print(balanced_df['isFraud'].value_counts())

# Guardamos el dataset balanceado en un archivo CSV
balanced_df.to_csv('balanced_dataset.csv', index=False)

Conteo de valores en 'isFraud' en el dataset balanceado:
isFraud
0    41326
1    20663
Name: count, dtype: int64


Podemos ver que ahora nuestro archivo "balanced_dataset.csv" cuenta con 41,326 "casos de no fraude" y 20,663 "casos de fraude". Ahora estamos listos para hacer el análisis exploratorio de nuestros datos.